### PR TITLE
Introduce Lua Extension: RocksLuaCompactionFilter

### PR DESCRIPTION
**Pull request by yhchiang**
Summary:
This diff includes an implementation of CompactionFilter that allows
users to write CompactionFilter in Lua.  With this ability, users can
dynamically change compaction filter logic without requiring building
the rocksdb binary and restarting the database.

To compile, WITH_LUA_PATH must be specified to the base directory
of lua.

Test Plan:
WITH_LUA_PATH=/path/to/lua make lua_test
./lua_test
make clean
make lua_test
./lua_test

### CODE REVIEW

**Comment by yhchiang** [comment]
The test failure of DBIteratorTest.PinnedDataIteratorRandomized on travis is not related to this PR:
https://travis-ci.org/facebook/rocksdb/jobs/174380511.

And the debian one seems not related to this PR:
https://travis-ci.org/facebook/rocksdb/jobs/174380507

**Comment by IslamAbdelRahman** [comment]
@yhchiang, looks like you need to rebase this PR

**Comment by IslamAbdelRahman** [comment]
@yhchiang, can you also re-import your diff after rebasing ?

**Comment by yhchiang** [comment]
Sure!

**Comment by yhchiang** [comment]
Rebased, and also address issues on Makefile and Lua library loading in certain Lua versions.

File: HISTORY.md [hunk 1]
```diff
@@ -7,6 +7,8 @@
 ### New Features
 * Add avoid_flush_during_shutdown option, which speeds up DB shutdown by not flushing unpersisted data (i.e. with disableWAL = true). Unpersisted data will be lost. The options is dynamically changeable via SetDBOptions().
 * Add memtable_insert_with_hint_prefix_extractor option. The option is mean to reduce CPU usage for inserting keys into memtable, if keys can be group by prefix and insert for each prefix are sequential or almost sequential. See include/rocksdb/options.h for more details.
+* Add avoid_flush_during_shutdown option, which speeds up DB shutdown by not flushing unpersisted data (i.e. with disableWAL = true). Unpersisted data will be lost. The options is dynamically changeable.
```
Comment (IslamAbdelRahman) [inline HISTORY.md]: Let's fix that

File: utilities/lua/rocks_lua_compaction_filter.cc [hunk 2]
```diff
@@ -0,0 +1,240 @@
+//  Copyright (c) 2016, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under the BSD-style license found in the
+//  LICENSE file in the root directory of this source tree. An additional grant
+//  of patent rights can be found in the PATENTS file in the same directory.
+
+#if defined(LUA) && !defined(ROCKSDB_LITE)
```
Comment (IslamAbdelRahman) [inline utilities/lua/rocks_lua_compaction_filter.cc]: Why do we need to do that ? should not we disable LUA if ROCKSDB_LITE is enabled ?

File: build_tools/build_detect_platform [hunk 3]
```diff
@@ -466,3 +466,4 @@ echo "PROFILING_FLAGS=$PROFILING_FLAGS" >> "$OUTPUT"
 if test -n "$JEMALLOC"; then
   echo "JEMALLOC=1" >> "$OUTPUT"
 fi
+echo "WITH_LUA_PATH=$WITH_LUA_PATH" >> "$OUTPUT"
```
Comment (IslamAbdelRahman) [inline build_tools/build_detect_platform]: Looks like this is not correct, right ?

File: HISTORY.md (hunk 1 above)
Comment (yhchiang) [inline HISTORY.md]: Looks like a bad merge.  Let me fix it.

File: utilities/lua/rocks_lua_compaction_filter.cc (hunk 2 above)
Comment (yhchiang) [inline utilities/lua/rocks_lua_compaction_filter.cc]: All utilities are not supported in RocksDBLite.

Comment (IslamAbdelRahman) [inline utilities/lua/rocks_lua_compaction_filter.cc]: btw, what I am trying to say here, is that when LITE is enabled, LUA is disabled. So it looks to me that here we can only check if LUA is enabled

**Pull request closed without merging**